File: cantorset.c
#include <stdio.h>
#define W 81
#define H 5

typedef unsigned int uint;
char matrix[100][100];

void init() {

     for(uint i = 0; i < H; ++i) {

     	 for(uint j = 0; j < W; ++j) {

     	 	 matrix[i][j] = '*';
     	 }
     }
}

void cantor(int start, int width, int index) {

     int segment = width / 3;

     if(segment == 0) return;

     for(uint i = index; i < H; ++i) {

     	 for(uint j = start + segment; j < start + 2 * segment; ++j) {

     	 	 matrix[i][j] = ' ';
     	 }
     }

     cantor(start, segment, index + 1);   
     cantor(start + 2 * segment, segment, index + 1);
       
}

void display() {

 for(uint i = 0; i <= H; ++i) {

     	 for(uint j = 0; j <= W; ++j) {

     	 	 printf("%c", matrix[i][j]);
     	 }
        	 printf("\n");
     } 	
}

int main(int argc, char const *argv[])
{
	init(); 
	cantor(0, W, 1);		
	display();
	return 0;
}

File: cantorset.go
package main

import "fmt"

const (	
	  width = 81
	  height = 5
)

var matrix[height][width] byte

func init() {

	 for i:=0; i < height; i++ {

	 	 for j:= 0; j < width; j++ {

	 	 	 matrix[i][j] = '*'
	 	 }
	 }
}

func display() {

	 for _, line := range matrix {

	 	  fmt.Println(string(line[:]))
	 }
}

func cantor(start, len, index int) {

	seg := len / 3

	if seg == 0 {

	   return	
	}

	for i := index; i < height; i++ {

		for j := start + seg; j < start + 2 * seg; j++  {

			matrix[i][j] = ' '
		} 
	}
    
    cantor(start, seg, index + 1)
    
    cantor(start + seg * 2, seg, index + 1) 

}

func main() {

    cantor(0, width, 1) 

	display()

}
